lib/user_space/user_x86_64_atomic.h

```diff
@@ -56,122 +56,170 @@ static inline int64_t m0_atomic64_get(const struct m0_atomic64 *a)
 }
 
 /**
- atomically increment counter
-
- @param a pointer to atomic counter
+   Atomically adds given amount to a counter
+ */
+static inline void m0_atomic64_add(struct m0_atomic64 *v, int64_t num)                    
+{                                                                       
+        long result;                                                    
+        unsigned long tmp;
+	asm volatile("// atomic64_add \n"			\
+	"       prfm    pstl1strm, %2\n"			\
+	"1:     ldxr    %0, %2\n"				\
+	"       add     %0, %0, %3\n"				\
+	"       stxr    %w1, %0, %2\n"                          \
+	"       cbnz    %w1, 1b"				\
+        : "=&r" (result), "=&r" (tmp), "+Q" (v->a_value)	\
+        : "Ir" (num));  
+                                                  
+}
 
- @return none
+/**
+   Atomically subtracts given amount from a counter
  */
-static inline void m0_atomic64_inc(struct m0_atomic64 *a)
-{
-	asm volatile("lock incq %0"
-		     : "=m" (a->a_value)
-		     : "m" (a->a_value));
+static inline void m0_atomic64_sub(struct m0_atomic64 *v, int64_t num)                    
```

Comment:
Different version of these functions are needed for X86 or ARM.

Reply:
Modified accordingly with conditional macro